src/core/vectortile/qgsvectortilemvtdecoder.cpp

```diff
@@ -0,0 +1,322 @@
+
+      for ( int tagNum = 0; tagNum < feature.tags_size(); tagNum += 2 )
+      {
+        int keyIndex = static_cast<int>( feature.tags( tagNum ) );
+        int fieldIndex = tagKeyIndexToFieldIndex.value( keyIndex, -1 );
+        if ( fieldIndex == -1 )
+          continue;
+
+        int valueIndex = static_cast<int>( feature.tags( tagNum + 1 ) );
+        const ::vector_tile::Tile_Value &value = layer.values( valueIndex );
+
+        if ( value.has_string_value() )
+          f.setAttribute( fieldIndex, QString::fromStdString( value.string_value() ) );
+        else if ( value.has_float_value() )
+          f.setAttribute( fieldIndex, static_cast<double>( value.float_value() ) );
+        else if ( value.has_double_value() )
+          f.setAttribute( fieldIndex, value.double_value() );
+        else if ( value.has_int_value() )
+          f.setAttribute( fieldIndex, static_cast<int>( value.int_value() ) );
+        else if ( value.has_uint_value() )
+          f.setAttribute( fieldIndex, static_cast<int>( value.uint_value() ) );
+        else if ( value.has_sint_value() )
+          f.setAttribute( fieldIndex, static_cast<int>( value.sint_value() ) );
+        else if ( value.has_bool_value() )
+          f.setAttribute( fieldIndex, static_cast<int>( value.bool_value() ) );  // or keep it bool? (do we have good support for that?)
+        else
+        {
+          // TODO: report - should not happen
+          Q_ASSERT( false );
+        }
+      }
+
+      //
+      // parse geometry
+      //
+
+      int extent = static_cast<int>( layer.extent() );
+      int cursorx = 0, cursory = 0;
+
+      QVector<QgsPoint *> outputPoints; // for point/multi-point
+      QVector<QgsLineString *> outputLinestrings;  // for linestring/multi-linestring
+      QVector<QgsPolygon *> outputPolygons;
+      QVector<QgsPoint> tmpPoints;
+
+      for ( int i = 0; i < feature.geometry_size(); i ++ )
+      {
+        unsigned g = feature.geometry( i );
+        unsigned cmdId = g & 0x7;
+        unsigned cmdCount = g >> 3;
+        if ( cmdId == 1 ) // MoveTo
+        {
+          for ( unsigned j = 0; j < cmdCount; j++ )
+          {
+            unsigned v = feature.geometry( i + 1 );
+            unsigned w = feature.geometry( i + 2 );
+            int dx = ( ( v >> 1 ) ^ ( -( v & 1 ) ) );
+            int dy = ( ( w >> 1 ) ^ ( -( w & 1 ) ) );
+            cursorx += dx;
+            cursory += dy;
+            double px = tileXMin + tileDX * double( cursorx ) / double( extent );
+            double py = tileYMax - tileDY * double( cursory ) / double( extent );
+
+            if ( feature.type() == vector_tile::Tile_GeomType_POINT )
+            {
+              outputPoints.append( new QgsPoint( px, py ) );
+            }
+            else if ( feature.type() == vector_tile::Tile_GeomType_LINESTRING )
+            {
+              if ( tmpPoints.size() > 0 )
+              {
+                outputLinestrings.append( new QgsLineString( tmpPoints ) );
```

Comment:
Note that if you're after maximum speed, the fastest like string constructor is the one which accepts seperate double vectors for X and y

Reply:
Good point... but from some timing test I did the decoding phase is quite fast - it is actually evaluation of filter expressions that's the slowest bit. If you don't mind I would leave this unchanged. (+ I find that separation of vectors for X and Y in QgsLineString is very strange - it should be more cache friendly if they were together, and just one array instead of two.)

Reply:
Mmm possibly? But it'd need to be an array of doubles and not QgsPoint (don't want to store all those unused z/m values). And then you'd need to upgrade all the existing performance sensitive code which uses the two vector constructor... 😜

Reply:
Yeah an array of doubles is what I had in mind too... and yes, it would be quite some work to get all that converted... maybe one day :-)